pkg/datapath/loader/compile.go

```diff
@@ -0,0 +1,182 @@
+		fmt.Sprintf("-D__NR_CPUS__=%d", runtime.NumCPU()),
+		"-Wno-address-of-packed-member", "-Wno-unknown-warning-option"}
+	standardLDFlags = []string{"-march=bpf", "-mcpu=probe"}
+
+	endpointPrefix   = "bpf_lxc"
+	endpointProg     = fmt.Sprintf("%s.%s", endpointPrefix, outputSource)
+	endpointObj      = fmt.Sprintf("%s.%s", endpointPrefix, outputObject)
+	endpointObjDebug = fmt.Sprintf("%s.dbg.%s", endpointPrefix, outputObject)
+	endpointAsm      = fmt.Sprintf("%s.%s", endpointPrefix, outputAssembly)
+
+	debugProgs = []*progInfo{
+		{
+			Source:     endpointProg,
+			Output:     endpointObjDebug,
+			OutputType: outputObject,
+		},
+		{
+			Source:     endpointProg,
+			Output:     endpointAsm,
+			OutputType: outputAssembly,
+		},
+		{
+			Source:     endpointProg,
+			Output:     endpointProg,
+			OutputType: outputSource,
+		},
+	}
+	datapathProg = &progInfo{
+		Source:     endpointProg,
+		Output:     endpointObj,
+		OutputType: outputObject,
+	}
+)
+
+// irPath determines a temporary path on the filesystem to store the compiled
+// intermediate representation object for the program.
+//
+// TODO: Remove this and just pipe the 'clang' and 'llc' programs together.
+func irPath(prog *progInfo, dir *directoryInfo) string {
+	return fmt.Sprintf("%s/%s.bc", dir.Output, prog.Output)
+}
+
+// progLDFlags determines the loader flags for the specified prog and paths.
+func progLDFlags(prog *progInfo, dir *directoryInfo) []string {
+	return []string{
+		fmt.Sprintf("-filetype=%s", prog.OutputType),
+		irPath(prog, dir),
+		"-o", fmt.Sprintf("%s/%s", dir.Output, prog.Output),
+	}
+}
+
+// Link the specified program from the specified path to the intermediate
+// representation, to the output specified in the prog's info.
+func Link(ctx context.Context, prog *progInfo, dir *directoryInfo, debug bool) error {
+	args := make([]string, 0, 8)
+	if debug {
+		args = append(args, "-mattr=dwarfris")
+	}
+	args = append(args, standardLDFlags...)
+	args = append(args, progLDFlags(prog, dir)...)
+
+	out, err := exec.CommandContext(ctx, linker, args...).CombinedOutput()
+	if err != nil {
+		scanner := bufio.NewScanner(bytes.NewReader(out))
+		return fmt.Errorf("Failed to link %q: %s: %q", prog.Output, err, scanner.Text())
+	}
+
+	return nil
+}
+
+// progLDFlags determines the compiler flags for the specified prog and paths.
+func progCFlags(prog *progInfo, dir *directoryInfo) []string {
+	var output string
+
+	if prog.OutputType == outputSource {
+		output = fmt.Sprintf("%s/%s", dir.Output, prog.Output)
+	} else {
+		output = irPath(prog, dir)
+	}
+
+	return []string{
+		fmt.Sprintf("-I%s/globals", dir.Runtime),
+		fmt.Sprintf("-I%s", dir.Output),
+		fmt.Sprintf("-I%s/include", dir.Library),
+		"-c", fmt.Sprintf("%s/%s", dir.Library, prog.Source),
+		"-o", output,
+	}
+}
+
+// compile and link a program.
+func compile(ctx context.Context, prog *progInfo, dir *directoryInfo, debug bool) error {
+	args := make([]string, 0, 16)
```

Comment:
why 16?

Reply:
Similar to ldflags, I looked at how many cflags we use.